sgx/trusted/src/wrapper/sgx_retrieve_key_impl.c

```diff
@@ -212,9 +412,100 @@ int enclave_retrieve_key(EVP_PKEY * enclave_sign_privkey, X509 * peer_cert)
+    close_socket_ocall(socket_fd);
+    return EXIT_FAILURE;
+  }
+
+  unsigned char *encrypted_request = NULL;
+  size_t encrypted_request_len = 0;
+
+  ret_val = aes_gcm_encrypt(session_key, session_key_len,
+                            key_request, key_request_len,
+                            &encrypted_request, &encrypted_request_len);
+  kmyth_enclave_clear_and_free(key_request, key_request_len);
+  if (ret_val)
+  {
+    kmyth_sgx_log(LOG_ERR, "Failed to encrypt the KMIP key request.");
+    kmip_destroy(&kmip_context);
+    kmyth_enclave_clear_and_free(session_key, session_key_len);
+    close_socket_ocall(socket_fd);
+    return EXIT_FAILURE;
+  }
+
+  // send request and get response from key server
+  unsigned char *encrypted_response = NULL;
+  size_t encrypted_response_len = 0;
+
+  ret_ocall = retrieve_key_ocall(&ret_val,
+                                 encrypted_request,
+                                 encrypted_request_len,
+                                 &encrypted_response,
+                                 &encrypted_response_len, socket_fd);
+  kmyth_enclave_clear_and_free(encrypted_request, encrypted_request_len);
+  close_socket_ocall(socket_fd);
+  if (ret_ocall != SGX_SUCCESS || ret_val != EXIT_SUCCESS)
+  {
+    kmyth_sgx_log(LOG_ERR, "Failed to send the KMIP key request.");
+    kmip_destroy(&kmip_context);
+    kmyth_enclave_clear_and_free(session_key, session_key_len);
+    return EXIT_FAILURE;
+  }
+
+  // decrypt response message
+  unsigned char *response = NULL;
+  size_t response_len = 0;
+
+  ret_val = aes_gcm_decrypt(session_key, session_key_len,
+                            encrypted_response, encrypted_response_len,
+                            &response, &response_len);
+  OPENSSL_free_ocall((void **) &encrypted_response);
   kmyth_enclave_clear_and_free(session_key, session_key_len);
+  if (ret_val)
+  {
+    kmyth_sgx_log(LOG_ERR, "Failed to decrypt the KMIP key response.");
+    kmip_destroy(&kmip_context);
+    return EXIT_FAILURE;
+  }
+
+  unsigned char *received_key_id = NULL, *key = NULL;
+  size_t received_key_id_len = 0, key_len = 0;
+
+  ret_val = parse_kmip_get_response(&kmip_context,
+                                    response, response_len,
+                                    &received_key_id, &received_key_id_len,
+                                    &key, &key_len);
+  kmyth_enclave_clear_and_free(response, response_len);
+  kmip_destroy(&kmip_context);
+  if (ret_val)
+  {
+    kmyth_sgx_log(LOG_ERR, "Failed to parse the KMIP Get response.");
+    return EXIT_FAILURE;
+  }
+
+  snprintf(msg, MAX_LOG_MSG_LEN, "Received a KMIP object with ID: %.*s",
+           (int) received_key_id_len, received_key_id);
+  kmyth_sgx_log(LOG_DEBUG, msg);
+
+  snprintf(msg, MAX_LOG_MSG_LEN, "Received operational key: 0x%02X..%02X",
+           key[0], key[key_len - 1]);
+  kmyth_sgx_log(LOG_DEBUG, msg);
+
   kmyth_sgx_log(LOG_DEBUG, "completed ECDH exchange");
```

Comment:
Now that the demo is more than the ECDH key agreement, this log message should probably be updated to something like "completed key retrieval from server into enclave"

Reply:
Yes, that's better.